### DISCUSSION TITLE
Adapting a Windows .bat script with @call vcvars64.bat

### DISCUSSION THREAD

**Discussion by nigels-com** [casey/just#1484] [Q&A]
I'm interested in using just as an alternative to a Windows .bat script.
The first step of the script is to call a vendor-provided script that sets various environment variables:

```
setlocal
@call  "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Auxiliary\Build\vcvars64.bat"
``` 

My understanding is that this step is necessary for the subsequent cmake invocations to detect and use the appropriate C/C++ toolchain.  The rest converts to a cross-platform set of just recipes:

```
# cmake configuration step
cmake:
	cmake -S . -B {{BUILD}} \
		{{ if os_family() == 'windows' { '-A x64'} else { '' } }} \
		-DCMAKE_INSTALL_PREFIX={{CMAKE_INSTALL_PREFIX}} \
		-DCMAKE_BUILD_TYPE={{CMAKE_BUILD_TYPE}} \
		{{CMAKE_FLAGS}}

# cmake build step
build:
	cmake --build {{BUILD}} -j

# installation step
install:
	cmake --build {{BUILD}} -j --target {{INSTALL}}

# packaging step
package:
	cmake --build {{BUILD}} -j --target dist
```

But it's not at all clear how I can use the environment variables via vcvars64.bat using just.

Edit: I'm expecting to use cmd.exe for windows-shell, but would be happy to have a powershell solution.

Thanks in advance for any suggestions.

**Comment by casey** [comment]
This has come up a few times before, in different contexts, and I've never been able to figure out how to support this pattern well. Just runs every line of every recipe in a fresh shell, so environment variables aren't persisted across invocations.

**Reply by nigels-com** [reply]
I think in our specific setup we could possibly move this "up" to the Jenkinsfile pipeline with some awkward cmd.exe invocation of just.
My team may well push back on that idea, but may well accept it as a good price to pay for cross-platform build script, and no more batch or GNU make wrangling.

**Reply by nigels-com** [reply]
It does make me think of cmd.exe /k or bash --rcfile as something that just could possibly support doing, but alas I'm not sure I have the opportunity to learn enough rust to take a stab at that.

**Reply by casey** [reply]
Thinking about it now, is there any way that you can write the env variables that the bat file creates to a .env file? If so, `just` supports loading .env files and exporting variables to recipes.

**Reply by nigels-com** [reply]
Yes, I did look into that to discover the true horror of various batch files that MS Visual Studio is using to populate those environment variables.  It would work here and now to bake them into an .env, but does seem fragile in terms of adding a newer toolchain, or whatnot.  MS do specifically advise against that specifically.  Some have gone to the trouble to unpack them, but I think that would be maintenance burden in our case.
https://renenyffenegger.ch/notes/Windows/development/Visual-Studio/environment-variables/index

I would say that despite this particular issue, I'm quite pleased with the way just slots into this particular niche - doing things in the make sort of way without the historical baggage, and happily without the necessity of tabs!  We definitely don't want to run cmake directly, and this codifies various things very nicely.

**Reply by nigels-com** [reply]
If I never need to write or debug another batch file, you've made me very happy indeed.

**Comment by nigels-com** [comment]
The workaround I've proposed to our team is to move the `vcvars64.bat` up the stack into the Jenkinsfile CI script.
This means that Windows devs need to source that into their environment, rather than have it handled by the batch file as it previously was.
But I think this is workable for us.

```
    stages {
        stage('Build') {
            steps {
                bat '@call "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Professional\\VC\\Auxiliary\\Build\\vcvars64.bat" && just all'
            }
        }

```

**Comment by runeimp** [comment]
Couldn't you shebang the invocation? Just does some shebang magic on Windows for PowerShell shebang scripting. Is that also true for CMD @casey?

**Reply by nigels-com** [reply]
Our intention is to use Just as a simplified cross-platform make alternative.  But since just isn't `cmd.exe` I've mostly come around to solving this "upstream" of Just.

**Comment by jkelin** [comment]
Since this comes up as one of the first results in google search, here is one way to do it:

```just
vsmake:
  #!cmd.exe /c
  call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"
  ./build.bat
```

for example to build a project with premake

```just
no_gfx_api_vma:
  just premake "no_gfx_api/gpu/vma" "--vk-version=3" "build"

premake dir args mode:
  #!cmd.exe /c
  call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"
  cd {{dir}}
  premake5 {{args}} vs2022
  cd build
  ./build.bat {{mode}}
```

Edit:
If you want to use powershell as well

```just
[windows]
[script("cmd.exe", "/c")]
[private]
vcvars *args:
  call "{{ `vswhere -property resolvedInstallationPath` / "VC" / "Auxiliary" / "Build" / "vcvars64.bat" }}"
  echo {{args}}
  cmd.exe /c {{ args }}

[windows]
[private]
vcvars-script path:
  cp "{{ path }}" "{{ path }}.ps1"
  just vcvars pwsh.exe -NoLogo -File "{{ path }}.ps1"

[windows]
msbuild-command:
  # example of running a single command with visual studio environment variables
  just vcvars msbuild --version

[windows]
[script("just", "vcvars-script")]
msbuild:
  # example of running a whole script with powershell and visual environment variables
  msbuild --version
```
